### ISSUE TITLE
heapdump not triggering on OOM

### ISSUE THREAD

**Issue by jdirwin** [blueconic/node-oom-heapdump#1]
Using version 1.0.11 on node 6.12.0; node-oom-heapdump started at server boot with only path option.
Got confirmation that it started:
  Debugger is listening on port 9229, 'node-oom-heapdump' can function correctly.

But when a OOM occurs, node-oom-heapdump isn't triggered and no dump file is written.

Console output from OOM attached.
```
<--- Last few GCs --->

326386181 ms: Mark-sweep 1322.2 (1436.6) -> 1322.2 (1436.6) MB, 1091.8 / 0.5 ms [allocation failure] [GC in old space requested].
326387262 ms: Mark-sweep 1322.2 (1436.6) -> 1322.2 (1436.6) MB, 1080.8 / 0.5 ms [allocation failure] [GC in old space requested].
326388400 ms: Mark-sweep 1322.2 (1436.6) -> 1331.4 (1420.6) MB, 1138.3 / 1.0 ms [last resort gc].
326389575 ms: Mark-sweep 1331.4 (1420.6) -> 1340.6 (1420.6) MB, 1174.8 / 1.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x363367acf781 <JS Object>
    1: set [/srv/apinode/node_modules/bookshelf/lib/base/model.js:~189] [pc=0x394ff81b7535] (this=0x26f3563a4069 <JS Object>,key=0x372aaa99a6d1 <an Object with map 0x26b5a7d84f79>,val=0x19e2c84a8819 <an Object with map 0x26de185605b9>,op
tions=0x363367a04381 <undefined>)
    2: arguments adaptor frame: 2->3
    3: constructor(aka ModelBase) [/srv/apinode/node_modules/bookshelf/lib/base/model.js:...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: 0x10a04bc [/usr/local/bin/node]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [/usr/local/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 5: v8::internal::Factory::NewTransitionArray(int) [/usr/local/bin/node]
 6: v8::internal::TransitionArray::Insert(v8::internal::Handle<v8::internal::Map>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Map>, v8::internal::SimpleTransitionFlag) [/usr/local/bin/node]
 7: v8::internal::Map::CopyReplaceDescriptors(v8::internal::Handle<v8::internal::Map>, v8::internal::Handle<v8::internal::DescriptorArray>, v8::internal::Handle<v8::internal::LayoutDescriptor>, v8::internal::TransitionFlag, v8::internal:
:MaybeHandle<v8::internal::Name>, char const*, v8::internal::SimpleTransitionFlag) [/usr/local/bin/node]
 8: v8::internal::Map::CopyAddDescriptor(v8::internal::Handle<v8::internal::Map>, v8::internal::Descriptor*, v8::internal::TransitionFlag) [/usr/local/bin/node]
 9: v8::internal::Map::CopyWithField(v8::internal::Handle<v8::internal::Map>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::FieldType>, v8::internal::PropertyAttributes, v8::internal::Representation, v8::in
ternal::TransitionFlag) [/usr/local/bin/node]
10: v8::internal::Map::TransitionToDataProperty(v8::internal::Handle<v8::internal::Map>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::StoreF
romKeyed) [/usr/local/bin/node]
11: v8::internal::LookupIterator::PrepareTransitionToDataProperty(v8::internal::Handle<v8::internal::JSObject>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::StoreFromKeyed) [/usr/loc
al/bin/node]
12: v8::internal::StoreIC::LookupForWrite(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::Object::StoreFromKeyed) [/usr/local/bin/node]
13: v8::internal::StoreIC::UpdateCaches(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::Object::StoreFromKeyed) [/usr/local/bin/node]
14: v8::internal::StoreIC::Store(v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::Object::StoreFromKeyed) [/usr/local/bin/node]
15: v8::internal::KeyedStoreIC::Store(v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [/usr/local/bin/node]
16: v8::internal::Runtime_KeyedStoreIC_Miss(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
17: 0x394febd092a7
```

**Comment by jdirwin** [comment]
Oh.  I guess this is one of the side effects of your issue https://github.com/node-inspector/v8-profiler/issues/109

Heap growth rate for the current approach has some limit, beyond which a fix for 109 is the only solution.

Our case is probably a single request making a huge allocation.

You might want to add a section to the Readme explaining this limitation.

Thanks for your contribution in any case, we'll keep using it in case of slower OOMs.

**Comment by paulrutter** [comment]
You're right that one huge allocation might not lead to an OoM heapdump, because the current logic only kicks in after a full gc. When there is no time for a full gc, this module won't create a heapdump.

Do notice that this module is not based on v8-profiler. But, the issue is still similar: nodejs doesn't offer a hook before an OoM occurs. That's why the current approach is needed, to simulate this behavior. This indeed is not a perfect solution, as you encountered. 

I'll add a limitation section to the readme, explaining the limitations off the current solution. If you have suggestions for improving on the current approach, let me know!

**Comment by paulrutter** [comment]
Ps. Do you start your nodejs process with the max_old_space_size, optimize_for_size,  always_compact flags?
The first flag limits your process heap size, the last two have impact on garbage collection. These flags might impact whether a full gc is done (earlier).

**Comment by paulrutter** [comment]
See https://github.com/blueconic/node-oom-heapdump/blob/master/README.md#large-allocations--memory-growth-volatility.

**Comment by paulrutter** [comment]
This should be fixed with the latest 1.1.0 release :)
